### PR TITLE
Add Hill Tensor Material

### PR DESCRIPTION
**Pull request by SudiptaBiswas**
Creates a new material class for Hill tensor (similar to elasticity tensor material).

Refs #16016

### CODE REVIEW

**Comment by SudiptaBiswas** [comment]
First iteration for HillTesnorMaterial. Putting this up for design discussion. @recuero @dschwen

**Comment by SudiptaBiswas** [comment]
This is ready for review @recuero @dschwen

**Comment by dschwen** [comment]
If we expect the till tensor material to be stateful at any point we might just want to store the coefficients.

Then we also should look into block matrix eigen decomposition. The Hill tensor has a particularly simple structure. It is symmetric, the block off-diagonals are zero and the lower right block is diagonal.

File: modules/tensor_mechanics/doc/content/source/materials/ADHillTensor.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  H_{rot} = T_m H T_m^T
+\end{equation}
+ Here, $H_{rot}$ is the rotated hill tensor, $H$ is the initial Hill tensor and $T_m$ is the transformation matrix.
+
+The transformation matrix is formed assuming anti-clockwise rotation about z-, y-, and x-, axis in sequence such that
+\begin{equation}
+\label{eq:transformation_matrix}
+ T_m = T_x T_y T_z
+\end{equation}
+Here, T_x, T_y, T_z are the transformation matrices about x-, y- and z-axis, respectively:
```
Comment (recuero) [inline modules/tensor_mechanics/doc/content/source/materials/ADHillTensor.md:21]: Can we provide a reference to the transformation matrices?

File: modules/tensor_mechanics/doc/content/source/materials/ADHillTensor.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+             \cos{\theta}^2 & \sin{\theta}^2 & 0 & 0 & 0 & 2 \cos{\theta} \sin{\theta}\\
+             \sin{\theta}^2 & \cos{\theta}^2 & 0 & 0 & 0 & - 2 \cos{\theta} \sin{\theta}\\
+             0 & 0 & 1 & 0 & 0 & 0\\
+             0 & 0 & 0 & \cos{\theta}& \sin{\theta} & 0\\
+             0 & 0 & 0 & - \sin{\theta}& \cos{\theta} & 0\\
+             -\cos{\theta} \sin{\theta} & \cos{\theta} \sin{\theta} & 0 & 0 & 0 & \cos{\theta}^2-\sin{\theta}^2
+             \end{bmatrix}}
+\end{aligned}
+\end{equation}
+where, $\alpha$, $\beta$ and $\theta$ are the rotation angles. Note that, for reducing the computational cost, only the components required for the updated hill constants are computed from the rotated Hill tesnor.
```
Comment (recuero) [inline modules/tensor_mechanics/doc/content/source/materials/ADHillTensor.md:51]: Suggested change:

Note that, to reduce the computational cost, only the components required for the updated hill constants are computed from the rotated Hill tensor.

File: modules/tensor_mechanics/src/materials/ADHillPlasticityStressUpdate.C [hunk 3]
```diff
@@ -131,6 +101,7 @@ ADHillPlasticityStressUpdate::computeDeltaDerivatives(const ADReal & delta_gamma
 {
   omega_gamma = 0.0;
   sy_gamma = 0.0;
+  computeHillTensorEigenDecomposition(_hill_tensor[_qp]);
```
Comment (recuero) [inline modules/tensor_mechanics/src/materials/ADHillPlasticityStressUpdate.C:104]: Can we compute the Eigendecomposition once per time step?
Doing that will involve some approximation once we account for the kinematics of the rotation increment. But, at the same time, omitting some inner loop decompositions should not affect convergence and may help us with computational efficiency. As the tests are set up, probably the eigendecomposition can be moved to an outer layer.

File: modules/tensor_mechanics/src/materials/ADHillPlasticityStressUpdate.C [hunk 4]
```diff
@@ -171,6 +142,7 @@ ADHillPlasticityStressUpdate::computeResidual(const ADDenseVector & stress_dev,
   // If in elastic regime, just return
   if (_yield_condition <= 0.0)
     return 0.0;
+  computeHillTensorEigenDecomposition(_hill_tensor[_qp]);
```
Comment (recuero) [inline modules/tensor_mechanics/src/materials/ADHillPlasticityStressUpdate.C:145]: Same here

File: modules/tensor_mechanics/src/materials/ADHillPlasticityStressUpdate.C [hunk 5]
```diff
@@ -313,6 +285,8 @@ ADHillPlasticityStressUpdate::computeStressFinalize(
   if (_yield_condition <= 0.0)
     return;
 
+  computeHillTensorEigenDecomposition(_hill_tensor[_qp]);
```
Comment (recuero) [inline modules/tensor_mechanics/src/materials/ADHillPlasticityStressUpdate.C:288]: Same here and in other instances

**Review by recuero — COMMENTED** [review]
👍

**Comment by recuero** [comment]
> If we expect the till tensor material to be stateful at any point we might just want to store the coefficients. Then we also should look into block matrix eigen decomposition. The Hill tensor has a particularly simple structure. It is symmetric, the block off-diagonals are zero and the lower right block is diagonal.

@dschwen  That'd be of a lot of help computationally.

File: modules/tensor_mechanics/src/materials/ADHillPlasticityStressUpdate.C [hunk 6]
```diff
@@ -47,34 +42,10 @@ ADHillPlasticityStressUpdate::ADHillPlasticityStressUpdate(const InputParameters
     _hardening_slope(0.0),
     _yield_condition(1.0),
     _yield_stress(getParam<Real>("yield_stress")),
+    _hill_constants(getMaterialProperty<std::vector<Real>>(_base_name + "hill_constants")),
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ADHillPlasticityStressUpdate.C:45]:
```suggestion
    _hill_constants(getMaterialPropertyByName<std::vector<Real>>(_base_name + "hill_constants")),
```

**Comment by SudiptaBiswas** [comment]
> > If we expect the till tensor material to be stateful at any point we might just want to store the coefficients. Then we also should look into block matrix eigen decomposition. The Hill tensor has a particularly simple structure. It is symmetric, the block off-diagonals are zero and the lower right block is diagonal.
> 
> @dschwen That'd be of a lot of help computationally.

I will address this in a separate PR.

**Comment by recuero** [comment]
Did you push the `HillConstants` class?

**Comment by SudiptaBiswas** [comment]
> Did you push the `HillConstants` class?

Yeah, did.

**Review by recuero — APPROVED** [review]
This looks great to me.

When we make the Hill constants dependent with temperature, we'll need to make `HillConstants` truly AD to capture the dependency.

@dschwen please take a look

**Review by dschwen — APPROVED** [review]
Looks good to me

**Pull request merged**